### PR TITLE
feat: check that the manifests declaring a version agree on it

### PR DESCRIPTION
**Pull request by AugustusW**
Follows #21. Thank you for the two implementation notes; both shaped the design rather than being bolted on.

## What it does

`scripts/check_versions.py` reads every file that declares a version and fails if they disagree. `.github/workflows/version-consistency.yml` runs it on push and pull request.

Current output on `main`:

```
  .claude-plugin/plugin.json        0.4.0
  .codex-plugin/plugin.json         0.4.0
  skills/sepia/SKILL.md             0.4.0
  .agents/plugins/marketplace.json  (no version declared)
  .claude-plugin/marketplace.json   (no version declared)
  plugin.json                       (no version declared)
  skills/sepia-recreate/SKILL.md    (no version declared)
  skills/sepia-refactor/SKILL.md    (no version declared)
  skills/sepia-review/SKILL.md      (no version declared)
  skills/sepia-write/SKILL.md       (no version declared)

version check: OK, 3 declarations, all 0.4.0.
```

## On your first note: discovery, not pinned paths

Nothing is hardcoded to a path. Any file named `plugin.json` or `marketplace.json` is read wherever it sits, and every `SKILL.md`'s frontmatter is read for `metadata.version`. Both JSON shapes are covered: a top-level `version`, and a `version` on each entry of a `plugins` array, which is where a marketplace file would put one.

So the root `plugin.json` from #14 is already in scope. It declares no version today, which is a legitimate state and is listed rather than failed. If it gains one later it is checked from that moment, with nobody needing to remember this file exists. Same for either `marketplace.json`, and for the four wrapper skills.

Printing the files that declare nothing is deliberate: it is how a maintainer sees what the scan actually reached, so a manifest that discovery started missing would show up as an absence in that list rather than as silence.

One more case, since it is the way a checker like this usually rots: **discovery finding nothing is a failure, not a pass.** If the layout moves and this script stops matching anything, it says so instead of going green.

## On your second note: no new dependencies

Standard library only. The frontmatter is parsed with a line scan rather than PyYAML, and the workflow calls `python3` directly rather than running a setup action, so nothing is installed and no lockfile appears. Verified on Python 3.9.6 as well as 3.12.

## Verification

Each case run against a copy of this repository, exit status checked:

| Case | Result |
|---|---|
| `main` as it stands | 3 declarations, all `0.4.0`, exit 0 |
| `.codex-plugin/plugin.json` set to `0.3.0` | `FAILED, 2 different versions declared: 0.3.0, 0.4.0`, exit 1 |
| **root `plugin.json` given `0.5.0`** | `FAILED, 2 different versions declared: 0.4.0, 0.5.0`, exit 1 |
| root `plugin.json` given a matching `0.4.0` | passes, now 4 declarations |
| a version added to `marketplace.json`'s `plugins[0]` | reported as `.claude-plugin/marketplace.json → plugins[0]  0.9.9` |
| run against a tree with no manifests at all | fails rather than passing vacuously |

The third row is your fourth-manifest case, and the fourth row is the same file behaving correctly once it agrees.

## What this does not do

It compares declared versions to each other, not to a release, a tag, or a CHANGELOG. Those live outside the working tree and a repository check cannot see them. Worth naming, because a guard's real coverage is whatever it can observe, which is not the same as the checklist it came from.

Happy to adjust naming, output format, or the workflow triggers to taste.

### CODE REVIEW

**Comment by AugustusW** [comment]
Revised per review. Both findings were real, and both were exactly the shape this checker exists to prevent, so thank you for testing it the way its own targets would fail.

## The two logic holes

**A version field that vanishes or changes type now fails instead of counting as absent.**

- `version` present but not a non-empty string is an error: `"version": 0.5` reports `version is 0.5, not a non-empty string`, exit 1. Same for `""`, and for a `plugins[]` entry.
- The three files that declare the version today are a required set. Discovery still scans everything, but these must keep declaring, so deleting `.codex-plugin/plugin.json`'s field now reports `required to declare a version and does not`, exit 1. The failure message says to update `REQUIRED` in the same commit when the layout changes on purpose.

Replaying your two cases against the revised script: removing the codex `version` exits 1 (was 0), and `"version": 0.5` exits 1 (was 0).

**Frontmatter is scoped to `metadata:`.** The regex is gone, replaced by a line scan with one bit of state: entering the top-level `metadata:` key sets it, the next top-level key clears it, and only a `version:` inside counts. Your `compatibility.version: "9.9.9"` case now passes with the skill correctly read as `0.4.0` and `9.9.9` appearing nowhere in the report; it is also a test case, as is a top-level frontmatter `version:`, which is likewise ignored.

## The other notes

- **Formal tests:** the manual cases from the PR description are now `tests/test_check_versions.py`, fifteen `unittest` cases, standard library only. The workflow runs them before the checker itself, so the checker regressing is a red build rather than a quiet one.
- **Workflow hardening:** `permissions: contents: read` added, and `actions/checkout` pinned to the full commit SHA behind the `v4` tag (`11d5960a`, verified against `actions/checkout` before pinning). `workflow_dispatch:` added so it can be run by hand.
- **Remote CI actually exercised:** the workflow has now run for real on my fork's main, green on both steps: https://github.com/AugustusW/sepia/actions/runs/33511269761 (checks on this PR itself will stay pending until the workflow exists in the base repository, since a new workflow file arriving from a fork PR does not run there).
- **Attribution:** commits rewritten with my linked email; the `localdomain` author is gone.
- One small addition while here: the repository had no `.gitignore`, and this PR brings its first Python, so there is now one ignoring `__pycache__/`.

**Comment by AugustusW** [comment]
Fixed in c6e3264.

**Direct children only.** The block's first child fixes the indentation that direct children of `metadata:` sit at, and lines at any other depth are skipped. Your case now reads the skill as `0.4.0` with `9.9.9` appearing nowhere in the report, exit 0:

```yaml
metadata:
  compatibility:
    version: "9.9.9"
  version: "0.4.0"
```

Both orderings are covered, and a metadata block containing only nested versions declares none, which for a required file is a failure rather than a silent shrink.

**One adjacent bug found while fixing it:** a blank line inside the frontmatter was treated as a top-level key, which closed the metadata block, so a `version:` after the blank line was missed. Same class of defect you have been catching, found from the same angle: feed the parser shapes that are legal in the wild. Blank lines are now skipped.

All three cases are tests; the suite is at eighteen, green on 3.9 and 3.12, and the fork CI run on this SHA is green: https://github.com/AugustusW/sepia/actions/runs/33511971952

**Comment by Nanako0129** [comment]
@codex review

**Comment by Nanako0129** [comment]
@AugustusW Codex flagged one thing worth a decision from you, before or after merge, your call: the frontmatter scan doesn't recognize flow-style `metadata: {version: "0.4.0"}`, so a reformat to flow syntax would fail the REQUIRED check as "missing" rather than being read. Two answers are equally acceptable from our side: support the flow form, or keep the line scan and fail that shape explicitly as unsupported, which matches your own present-but-wrong-is-an-error rule and makes the message name the real cause. We are also fine merging as-is and taking it as a follow-up; the failure it produces today is loud and named, which is the property we care about most.

**Comment by AugustusW** [comment]
Fixed in 4d7fd39.

Correct catch: `metadata: {version: "0.4.0"}` never entered the block, so the required canonical skill was reported as not declaring, and a harmless formatter pass could fail CI. Fail-closed, but wrongly.

A flow mapping is now read with the same direct-child semantics as block style: innermost `{...}` groups are deleted until none remain, so a version inside `compatibility: {...}` is gone before the scan and only metadata's own level is searched. Two honest boundaries, both reported explicitly instead of guessed at: braces inside quoted string values do not reduce cleanly and are named as such, and a scalar `metadata:` value is named as neither block nor flow.

Four new tests (plain flow, nested flow keeping direct-child semantics, nested-only declaring nothing, scalar value); the suite is at 22, green on 3.9 and 3.12, fork CI green on this SHA: https://github.com/AugustusW/sepia/actions/runs/33518908865

**Comment by Nanako0129** [comment]
@codex review

**Comment by Nanako0129** [comment]
@AugustusW On the three new findings, from our side: the first two are straightforwardly right under your own present-but-wrong rule, nothing to decide. On the third, the stance we would take: do not implement YAML scalar typing. Require the value to be quoted or fail any unquoted value that YAML would read as a number, boolean, or null, with a message that says to quote it. That keeps the scanner a scanner.

Also noted for our own playbook that `@codex review` is what re-triggers the bot on a fork push. Thanks for driving your own review loop.

**Comment by AugustusW** [comment]
All three fixed in the latest push.

**Whitespace** is now invalid rather than normalized: `" 0.4.0 "` reports `not a non-empty string without surrounding whitespace` instead of quietly agreeing with `"0.4.0"`, and quoted frontmatter values are held to the same rule as the JSON manifests.

**Comments** carry no structure at any indentation. An unindented `#` line between `metadata:` and its children no longer closes the block, and an indented one no longer fixes the child-indentation level (that second case was adjacent: had the comment been treated as the first child, its depth would have become the required level and the real version would have been skipped).

**Unquoted scalars**, per your direction, keep the scanner a scanner: no YAML typing. An unquoted value that YAML would read as a number, boolean or null is rejected with `...quote it`; an unquoted `0.4.0` is not a valid YAML number and remains an accepted plain string, so existing unquoted dotted versions do not become false reds.

Both frontmatter styles now share one scalar rule (`_frontmatter_scalar`), so flow and block cannot drift apart on value handling. Eight new tests, one per finding plus the boundary cases above; the suite is at 30, green on 3.9 and 3.12, fork CI green on this SHA: https://github.com/AugustusW/sepia/actions/runs/33522279856

@codex review

**Comment by Nanako0129** [comment]
@codex review

**Comment by Nanako0129** [comment]
@AugustusW Round three, two findings, our take:

The path-parts one is a straight fix: test the skip names against the repository-relative parts only.

On the scalar regex: stop extending it. Our earlier suggestion to enumerate what YAML reads as non-strings is what produced this round, and the enumeration has no natural end (binary now, then inf and nan, then sexagesimal). Invert it: an unquoted metadata.version is invalid, full stop, with the message saying to quote it. The repo's own file already quotes its version, the check gets simpler, and this entire class of finding closes permanently.

**Comment by AugustusW** [comment]
Both fixed in the latest push.

**Inverted, as directed.** Any unquoted `metadata.version` is now invalid with `...is unquoted; quote it`, the scalar regex is deleted, and the docstring records why: enumerating what YAML reads as non-strings had no natural end, and each extension of the list was this class of finding regenerating. You are right that our earlier suggestion produced the round; the enumeration was answering "what does YAML do" when the only question that closes is "what do we require". The round-four test that accepted unquoted `0.4.0` is inverted rather than deleted, since leaving it would have locked the old behavior in with a green light.

**Skip dirs scoped below root.** `SKIP_DIRS` matched absolute path parts, so a checkout at `.../venv/<repo>` skipped every file and reported all required manifests missing. Only directories below `root` are consulted now, and a test pins a baseline repository nested under a `venv` ancestor.

Suite is at 29, green on 3.9 and 3.12, fork CI green on this SHA: https://github.com/AugustusW/sepia/actions/runs/33524739695

@codex review

**Comment by Nanako0129** [comment]
@codex review

**Comment by Nanako0129** [comment]
@AugustusW Round four, one finding, and it is the second one to live inside the flow-mapping parser, so our stance is the same subtraction that closed the scalar class: delete the flow parser and report flow-style `metadata:` as unsupported with a message that says to use block style. The failure a formatter would then cause is loud and names its own fix, the repo controls its own frontmatter style anyway, and the whole class of quoted-span and nesting edge cases closes at once. Net negative diff again, most likely.

If you would rather keep flow support, stripping quoted spans before the key scan is acceptable to us too, but every finding since round one has traced back to hand-parsing ambitions, and both closures that worked so far were deletions.

**Comment by AugustusW** [comment]
You are right, and plainly so: the earlier rounds were attachment to hand-parsing, and each fix was feeding an endless loop. Every finding since round one traced back to that ambition, and both closures that actually worked were deletions. Round three's flow support was an overreaction to a false red; the correct answer already existed then, which was to refuse loudly.

So this is the same subtraction. The flow parser is deleted. Any inline `metadata:` value, flow mapping or scalar, is refused with one message: `metadata is declared inline; flow-style metadata is not supported, use block style`. The failure a formatter causes is loud and names its own fix, and the whole class of quoted-span and nesting edge cases closes at once.

Net negative diff, as you predicted: **-56 lines** (+34/−90). The five flow tests are rewritten as one refusal test whose last case is this round's finding itself (`{description: "current, version: 9.9.9"}` no longer has its quoted text mistaken for a version key); the suite is at 25, green on 3.9 and 3.12, fork CI green on this SHA: https://github.com/AugustusW/sepia/actions/runs/33526453485

@codex review

**Comment by Nanako0129** [comment]
@codex review

**Comment by Nanako0129** [comment]
@codex review

**Comment by Nanako0129** [comment]
@AugustusW Round five, two findings, and a boundary decision from our side.

The delimiter one is a must: prefix-matching `---` lets a broken frontmatter keep the required-core guard green, which is the one promise this script exists to keep. Exact-match the delimiter lines.

The traversal pruning is performance, not correctness, and CI runs on clean checkouts. Fix it in the same push if it is quick for you, or we will take it as a follow-up issue; either answer is fine.

And the boundary: this round is the last we will process inside this PR. After your next push we trigger one closing review; any further P2 goes to a follow-up issue and we merge. Five rounds have verified the core promise from every direction, and a guard script does not need an unbounded review loop.

**Comment by AugustusW** [comment]
Fixed in the latest push. Delimiters are now whole lines that are exactly `---`: `---oops` no longer opens a block and `----` no longer closes one, so a broken declaration turns the required check red instead of leaving the scanner reading a version out of a frontmatter a real YAML loader would refuse.

One deliberate tolerance travels with the exactness: an editor-added UTF-8 BOM is stripped before the first line is compared, since real frontmatter loaders tolerate it and an invisible byte should not turn a valid required file red.

We also enumerated the adjacent edges before pushing and left them alone on purpose: CRLF works via `splitlines`; tab-indented metadata children go red, which matches real YAML loaders refusing tabs; a quoted value containing `" #"` mis-splits into a loud `quote it` error rather than a silent pass — handling it properly would mean tracking quoted spans, which is the hand-parsing ambition this PR just finished deleting. Every remaining edge fails loud and names a fix; none goes quietly green.

Two new delimiter tests plus the BOM case; suite is at 28, green on 3.9 and 3.12, fork CI green on this SHA: https://github.com/AugustusW/sepia/actions/runs/33527831859

@codex review

**Comment by Nanako0129** [comment]
@codex review

File: scripts/check_versions.py [hunk 1]
```diff
@@ -0,0 +1,305 @@
...
+        if not in_metadata:
+            continue
+        indent = len(line) - len(line.lstrip())
+        if child_indent is None:
+            child_indent = indent
+        if indent != child_indent:
+            continue
+        stripped = line.strip()
+        if stripped.startswith("version:"):
+            return _frontmatter_scalar(stripped[len("version:"):], rel)
```
[the remark this answers is not in the document]
Comment (Nanako0129) [reply scripts/check_versions.py:216]: Deferred to https://github.com/Nanako0129/sepia/issues/39 per the boundary set in the round-five comment: the closing round's findings go to follow-up issues and the PR merges. The direction there follows the script's own present-but-wrong rule (duplicate key = explicit invalid, not first-wins).

File: scripts/check_versions.py [hunk 2]
```diff
@@ -0,0 +1,241 @@
...
+        return None, []
+
+    in_metadata = False
+    child_indent = None
+    for line in text[3:end].splitlines():
+        if not line.strip():
+            continue
+        if not line[:1].isspace():
+            in_metadata = line.rstrip() == "metadata:"
+            child_indent = None
```
[the remark this answers is not in the document]
Comment (Nanako0129) [reply scripts/check_versions.py:205]: Addressed across 4d7fd39 and finally 4332c13: flow-style metadata is now refused with an explicit use-block-style message rather than parsed. Verified by mutation on our side.

File: scripts/check_versions.py [hunk 3]
```diff
@@ -0,0 +1,282 @@
...
+        data = json.loads(path.read_text(encoding="utf-8"))
+    except (OSError, json.JSONDecodeError) as exc:
+        return [], [(rel, f"unreadable: {exc}")]
+
+    declared, invalid = [], []
+    if isinstance(data, dict):
+        if "version" in data:
+            if _valid(data["version"]):
+                declared.append((rel, data["version"].strip()))
+            else:
```
[the remark this answers is not in the document]
Comment (Nanako0129) [reply scripts/check_versions.py:128]: Fixed in 3ee2f2f: surrounding whitespace is invalid rather than normalized. Mutation-verified.

File: scripts/check_versions.py [hunk 4]
```diff
@@ -0,0 +1,282 @@
...
+            if head.strip() == "metadata" and sep:
+                rest = rest.split(" #")[0].strip()
+                if rest:
+                    # Flow syntax: metadata: {version: "0.4.0", ...}. Valid
+                    # YAML, and a formatter may produce it, so a version here
+                    # must be found rather than reported missing.
+                    return _flow_metadata_version(rest, rel)
+                in_metadata = True
+            else:
+                in_metadata = False
```
[the remark this answers is not in the document]
Comment (Nanako0129) [reply scripts/check_versions.py:204]: Fixed in 3ee2f2f: comment-only lines carry no structure. Mutation-verified with an unindented comment inside the metadata block.

File: scripts/check_versions.py [hunk 5]
```diff
@@ -0,0 +1,282 @@
...
+        indent = len(line) - len(line.lstrip())
+        if child_indent is None:
+            child_indent = indent
+        if indent != child_indent:
+            continue
+        stripped = line.strip()
+        if stripped.startswith("version:"):
+            raw = stripped[len("version:"):].split(" #")[0].strip().strip("\"'")
+            if raw:
+                return raw, []
```
[the remark this answers is not in the document]
Comment (Nanako0129) [reply scripts/check_versions.py]: Closed by the quote inversion in d22086a: any unquoted metadata.version is invalid. Mutation-verified.

File: scripts/check_versions.py [hunk 6]
```diff
@@ -0,0 +1,326 @@
...
+    ".codex-plugin/plugin.json",
+    "skills/sepia/SKILL.md",
+)
+
+
+def _iter_files(root):
+    for path in sorted(root.rglob("*")):
+        if not path.is_file():
+            continue
+        if any(part in SKIP_DIRS for part in path.parts):
```
[the remark this answers is not in the document]
Comment (Nanako0129) [reply scripts/check_versions.py]: Fixed in d22086a: skip names are tested against repository-relative parts only. Verified by running the check from a /tmp/venv/sepia layout.

File: scripts/check_versions.py [hunk 7]
```diff
@@ -0,0 +1,326 @@
...
+    so it is reported instead of stripped (review round 4)."""
+    return isinstance(value, str) and value != "" and value == value.strip()
+
+
+# Unquoted scalars that YAML reads as something other than a string. Per the
+# maintainer's direction the scanner stays a scanner: no YAML typing is
+# implemented, these are simply rejected with instructions to quote the value.
+_YAML_NONSTRING_RE = re.compile(
+    r"^(?:[-+]?(?:\d+|\d*\.\d+|\d+\.\d*)(?:[eE][-+]?\d+)?"
+    r"|0x[0-9a-fA-F]+|0o[0-7]+"
```
[the remark this answers is not in the document]
Comment (Nanako0129) [reply scripts/check_versions.py]: Superseded by the quote inversion in d22086a: the scalar regex is gone entirely, so binary and every other unquoted form is invalid by rule. Mutation-verified with 0b10.

File: scripts/check_versions.py [hunk 8]
```diff
@@ -0,0 +1,313 @@
...
+    inner = rest[1:-1]
+    while True:
+        reduced = re.sub(r"\{[^{}]*\}", "", inner)
+        if reduced == inner:
+            break
+        inner = reduced
+    if "{" in inner or "}" in inner:
+        return None, [(rel, "metadata flow mapping has unbalanced braces; use block style")]
+
+    match = re.search(r"(?:^|,)\s*version:\s*([^,]*)", inner)
```
[the remark this answers is not in the document]
Comment (Nanako0129) [reply scripts/check_versions.py]: Closed in 4332c13 by deleting the flow parser: the quoted-span trap now yields the explicit unsupported error. Mutation-verified with the description-containing-version case.

File: scripts/check_versions.py [hunk 9]
```diff
@@ -0,0 +1,294 @@
...
+    than being mistaken for a top-level key.
+    """
+    try:
+        text = path.read_text(encoding="utf-8")
+    except OSError as exc:
+        return None, [(rel, f"unreadable: {exc}")]
+
+    if not text.startswith("---"):
+        return None, []
+    end = text.find("\n---", 3)
```
[the remark this answers is not in the document]
Comment (Nanako0129) [reply scripts/check_versions.py]: Fixed in a4cd8a6: delimiters must be exactly ---, with a deliberate BOM tolerance. Mutation-verified with ---oops and ---- variants.

File: scripts/check_versions.py [hunk 10]
```diff
@@ -0,0 +1,294 @@
...
+# purpose, change this list in the same commit.
+REQUIRED = (
+    ".claude-plugin/plugin.json",
+    ".codex-plugin/plugin.json",
+    "skills/sepia/SKILL.md",
+)
+
+
+def _iter_files(root):
+    for path in sorted(root.rglob("*")):
```
[the remark this answers is not in the document]
Comment (Nanako0129) [reply scripts/check_versions.py:65]: Deferred to https://github.com/Nanako0129/sepia/issues/36 per the review boundary: performance only, CI runs on clean checkouts.

**Comment by Nanako0129** [comment]
Merged, with the .gitignore add/add union resolved on our side. Thank you for seven rounds of this; the two subtraction commits are the ones we will be pointing people at.
